### PR TITLE
improve GraphicsWidget paint speed by caching the bounding rect and the path used by the shape method

### PR DESCRIPTION
**Pull request by danielhrisca**
By caching the bounding rect the painting is increased significantly. On a live updating plot with ~160 curves the timings show the following:

#### With this patch 0.95 us/call
image

#### Without 1.99 us/call
image

Overall the painting speed went from 43.5ms/frame to 39.3ms/frame

### CODE REVIEW

**Comment by danielhrisca** [comment]
I've also added the caching of the path used by the ``shape`` method.

With ``shape`` caching:

``shape`` takes 0.18 us/call and the painting takes 36ms/frame
image

Without ``shape caching
``shape`` takes 2.48 us/call and the painting takes 39ms/frame
image

File: pyqtgraph/graphicsItems/GraphicsWidget.py [hunk 1]
```diff
@@ -45,12 +49,25 @@ def width(self):
...
         return br
-        
+
     def shape(self):  ## No idea why this is necessary, but rotated items do not receive clicks otherwise.
-        p = QtGui.QPainterPath()
-        p.addRect(self.boundingRect())
-        #print "shape:", p.boundingRect()
+        p = self._path
+        if p is None:
+            p = QtGui.QPainterPath()
```
Comment (j9ac9k) [inline pyqtgraph/graphicsItems/GraphicsWidget.py:68]: can probably put this in one line...

```python
self._path = p = QtGui.QPainterPath()
```

**Comment by j9ac9k** [comment]
hi @danielhrisca 

Thanks for this PR and the profiling results.   I would ask you rename the variables to better reflect that they hold cached values.  This LGTM otherwise!

**Comment by danielhrisca** [comment]
> I would ask you rename the variables

Done

File: pyqtgraph/graphicsItems/GraphicsWidget.py (hunk 1 above)
Comment (danielhrisca) [reply pyqtgraph/graphicsItems/GraphicsWidget.py]: done

File: pyqtgraph/graphicsItems/GraphicsWidget.py [hunk 2]
```diff
@@ -45,12 +49,25 @@ def width(self):
...
+            p.addRect(self.boundingRect())
+            self._path = p
+
+        # print "shape:", p.boundingRect()
```
[the remark this answers is not in the document]
Comment (danielhrisca) [reply pyqtgraph/graphicsItems/GraphicsWidget.py]: done

**Comment by danielhrisca** [comment]
@j9ac9k can you trigger a rerun since there was a network problem that caused the failure ?

**Comment by j9ac9k** [comment]
LGTM, thanks @danielhrisca

**Pull request merged**